File: src/anywidget_graph/backends/neo4j.py
"""Neo4j database backend using the Python driver."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from anywidget_graph.converters.cypher import CypherConverter

if TYPE_CHECKING:
    from neo4j import Driver


class Neo4jBackend:
    """Backend for Neo4j database using the Python driver.

    This executes queries Python-side (not in browser).
    Use for server-side notebooks or when browser connectivity is limited.

    Parameters
    ----------
    driver : Driver | None
        Existing Neo4j driver instance.
    uri : str | None
        Neo4j connection URI (e.g., "neo4j://localhost:7687").
    auth : tuple[str, str] | None
        Authentication tuple (username, password).
    database : str
        Database name to use (default: "neo4j").

    Example
    -------
    >>> from anywidget_graph.backends import Neo4jBackend
    >>> backend = Neo4jBackend(
    ...     uri="neo4j://localhost:7687",
    ...     auth=("neo4j", "password"),
    ... )
    >>> nodes, edges = backend.execute("MATCH (n)-[r]->(m) RETURN n, r, m LIMIT 10")

    Or with an existing driver:

    >>> from neo4j import GraphDatabase
    >>> driver = GraphDatabase.driver(uri, auth=auth)
    >>> backend = Neo4jBackend(driver=driver)
    """

    def __init__(
        self,
        driver: Driver | None = None,
        uri: str | None = None,
        auth: tuple[str, str] | None = None,
        database: str = "neo4j",
    ) -> None:
        self._driver = driver
        self._uri = uri
        self._auth = auth
        self._database = database
        self._converter = CypherConverter()
        self._owns_driver = driver is None  # Track if we created the driver

    @property
    def query_language(self) -> str:
        """Return the default query language."""
        return "cypher"

    @property
    def driver(self) -> Driver:
        """Get or create the Neo4j driver."""
        if self._driver is None:
            if self._uri is None:
                msg = "Either driver or uri must be provided"
                raise ValueError(msg)
            from neo4j import GraphDatabase

            self._driver = GraphDatabase.driver(self._uri, auth=self._auth)
        return self._driver

    def execute(self, query: str) -> tuple[list[dict], list[dict]]:
        """Execute Cypher query and return (nodes, edges).

        Parameters
        ----------
        query : str
            Cypher query to execute.

        Returns
        -------
        tuple[list[dict], list[dict]]
            Tuple of (nodes, edges) lists.
        """
        with self.driver.session(database=self._database) as session:
            result = session.run(query)
            records = list(result)  # Consume within session

        data = self._converter.convert(records)
        return data["nodes"], data["edges"]

    def fetch_schema(self) -> tuple[list[dict], list[dict]]:
        """Fetch Neo4j schema.

        Returns
        -------
        tuple[list[dict], list[dict]]
            Tuple of (node_types, edge_types) with label/type info and properties.
        """
        node_types: list[dict[str, Any]] = []
        edge_types: list[dict[str, Any]] = []

        with self.driver.session(database=self._database) as session:
            # Get node labels
            labels_result = session.run("CALL db.labels()")
            for record in labels_result:
                label = record[0]
                # Get properties for this label
                props_result = session.run(f"MATCH (n:`{label}`) UNWIND keys(n) AS key RETURN DISTINCT key LIMIT 20")
                properties = [r[0] for r in props_result]
                node_types.append({"label": label, "properties": properties})

            # Get relationship types
            rel_types_result = session.run("CALL db.relationshipTypes()")
            for record in rel_types_result:
                rel_type = record[0]
                # Get properties for this type
                props_result = session.run(
                    f"MATCH ()-[r:`{rel_type}`]->() UNWIND keys(r) AS key RETURN DISTINCT key LIMIT 20"
                )
                properties = [r[0] for r in props_result]
                edge_types.append({"type": rel_type, "properties": properties})

        return node_types, edge_types

    def close(self) -> None:
        """Close the driver connection if we own it."""
        if self._driver and self._owns_driver:
            self._driver.close()
            self._driver = None

    def __enter__(self) -> Neo4jBackend:
        """Context manager entry."""
        return self

    def __exit__(self, *args: Any) -> None:
        """Context manager exit."""
        self.close()

File: src/anywidget_graph/ui/neo4j.js
/**
 * Neo4j database connection and query execution (browser-side).
 */
import neo4j from "https://cdn.jsdelivr.net/npm/neo4j-driver@5.28.0/lib/browser/neo4j-web.esm.min.js";

let driver = null;

/**
 * Connect to a Neo4j database.
 */
export async function connect(uri, username, password, model) {
  if (driver) {
    await driver.close();
    driver = null;
  }

  model.set("connection_status", "connecting");
  model.save_changes();

  try {
    driver = neo4j.driver(uri, neo4j.auth.basic(username, password));
    await driver.verifyConnectivity();
    model.set("connection_status", "connected");
    model.set("query_error", "");
    model.save_changes();

    await fetchSchema(model);
    return true;
  } catch (error) {
    model.set("connection_status", "error");
    model.set("query_error", "Connection failed: " + error.message);
    model.save_changes();
    driver = null;
    return false;
  }
}

/**
 * Disconnect from Neo4j database.
 */
export async function disconnect(model) {
  if (driver) {
    await driver.close();
    driver = null;
  }
  model.set("connection_status", "disconnected");
  model.save_changes();
}

/**
 * Check if connected to Neo4j.
 */
export function isConnected() {
  return driver !== null;
}

/**
 * Fetch database schema (labels and relationship types).
 */
export async function fetchSchema(model) {
  if (!driver) return;

  const database = model.get("connection_database") || "neo4j";
  const session = driver.session({ database });

  try {
    // Fetch node labels and their properties
    const labelsResult = await session.run("CALL db.labels()");
    const nodeTypes = [];

    for (const record of labelsResult.records) {
      const label = record.get(0);
      const propsResult = await session.run(
        `MATCH (n:\`${label}\`) UNWIND keys(n) AS key RETURN DISTINCT key LIMIT 20`
      );
      const properties = propsResult.records.map((r) => r.get(0));
      nodeTypes.push({ label, properties, count: null });
    }

    // Fetch relationship types and their properties
    const relTypesResult = await session.run("CALL db.relationshipTypes()");
    const edgeTypes = [];

    for (const record of relTypesResult.records) {
      const type = record.get(0);
      const propsResult = await session.run(
        `MATCH ()-[r:\`${type}\`]->() UNWIND keys(r) AS key RETURN DISTINCT key LIMIT 20`
      );
      const properties = propsResult.records.map((r) => r.get(0));
      edgeTypes.push({ type, properties, count: null });
    }

    model.set("schema_node_types", nodeTypes);
    model.set("schema_edge_types", edgeTypes);
    model.save_changes();
  } catch (error) {
    console.error("Failed to fetch schema:", error);
  } finally {
    await session.close();
  }
}

/**
 * Execute a Cypher query.
 */
export async function executeQuery(query, database, model) {
  if (!driver) {
    model.set("query_error", "Not connected to database");
    model.save_changes();
    return null;
  }

  model.set("query_running", true);
  model.set("query_error", "");
  model.save_changes();

  const session = driver.session({ database: database || "neo4j" });

  try {
    const result = await session.run(query);
    model.set("query_running", false);
    model.save_changes();
    return processRecords(result.records);
  } catch (error) {
    model.set("query_running", false);
    model.set("query_error", "Query error: " + error.message);
    model.save_changes();
    return null;
  } finally {
    await session.close();
  }
}

/**
 * Process Neo4j records into nodes and edges.
 */
function processRecords(records) {
  const nodes = new Map();
  const edges = [];

  for (const record of records) {
    for (const key of record.keys) {
      const value = record.get(key);
      processValue(value, nodes, edges);
    }
  }

  return { nodes: Array.from(nodes.values()), edges };
}

/**
 * Process a single Neo4j value (node, relationship, path, or array).
 */
function processValue(value, nodes, edges) {
  if (!value) return;

  if (neo4j.isNode(value)) {
    const nodeId = value.elementId || value.identity.toString();
    if (!nodes.has(nodeId)) {
      const props = {};
      for (const [k, v] of Object.entries(value.properties)) {
        props[k] = neo4j.isInt(v) ? v.toNumber() : v;
      }
      nodes.set(nodeId, {
        id: nodeId,
        label: props.name || props.title || value.labels[0] || nodeId,
        labels: value.labels,
        ...props,
      });
    }
  } else if (neo4j.isRelationship(value)) {
    const props = {};
    for (const [k, v] of Object.entries(value.properties)) {
      props[k] = neo4j.isInt(v) ? v.toNumber() : v;
    }
    edges.push({
      source: value.startNodeElementId || value.start.toString(),
      target: value.endNodeElementId || value.end.toString(),
      label: value.type,
      ...props,
    });
  } else if (neo4j.isPath(value)) {
    for (const segment of value.segments) {
      processValue(segment.start, nodes, edges);
      processValue(segment.end, nodes, edges);
      processValue(segment.relationship, nodes, edges);
    }
  } else if (Array.isArray(value)) {
    for (const item of value) {
      processValue(item, nodes, edges);
    }
  }
}
